src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -52,7 +53,8 @@ internal abstract partial class NamedTypeSymbol
 
             NamedTypeSymbol underlyingType = getTupleUnderlyingType(elementTypesWithAnnotations, syntax, compilation, diagnostics);
 
-            if (numElements >= ValueTupleRestPosition && diagnostics != null && !underlyingType.IsErrorType())
+            if (numElements >= ValueTupleRestPosition && diagnostics != null && !underlyingType.IsErrorType() &&
+                basesBeingResolved?.ContainsReference(underlyingType.OriginalDefinition) != true)
```

Comment:
>basesBeingResolved?.ContainsReference(underlyingType.OriginalDefinition) != true [](http://example.com/codeflow?start=16&length=80)

While I understand that this change prevents a cycle, the logic looks suspicious. Usually, a `basesBeingResolved` check leads to some form of a cancelation, an error is reported, or items in the list are getting skipped in some processing. This doesn't look like that, and it doesn't have the same effect. 
Also, I think that we are not fixing the root cause problem for the cycle, which, in my opinion, is the fact that we are trying to bind something while creating symbols for declared members in type. There were other instances of similar "too early" binding, but we got rid of them. The indexer name is probably the last one that we didn't address. Since we finally have a motivating test scenario, I think we should get rid of the binding by changing the naming scheme for indexers. There must be an alternative to using ```explicitInterfaceTypeOpt.ToDisplayString(SymbolDisplayFormat.ExplicitInterfaceImplementationFormat)``` for building a unique name. For example, we could simply use explicit interface specifier followed by the name identifier as it is spelled out in source, combined with the containing partial type declaration index.

Reply:
After offline discussion, filed an issue to track this: https://github.com/dotnet/roslyn/issues/70901.